File: nvim/lua/whence/record.lua
local M = {}

local host = require("whence.host")

local SECTION = {
  ["host/definition"] = "definition",
  ["host/references"] = "references",
  ["host/documentHighlight"] = "documentHighlight",
}

local active = nil

function M._rel(file, root)
  if root and root ~= "" then
    local rel = vim.fs.relpath(root, file)
    if rel then
      return rel
    end
  end
  return file
end

local function key(rec, params)
  return ("%s:%d:%d"):format(M._rel(params.file, rec.root), params.line, params.col)
end

local function relativised(rec, locations)
  local out = vim.deepcopy(locations)
  for _, loc in ipairs(out) do
    loc.file = M._rel(loc.file, rec.root)
  end
  return out
end

local function copy_source(rec, file, text)
  local rel = M._rel(file, rec.root)
  if rel == file then
    return
  end
  local path = rec.dir .. "/" .. rel
  vim.fn.mkdir(vim.fs.dirname(path), "p")
  -- "b" joins the lines with \n and adds no trailing newline, so this round-trips exactly.
  vim.fn.writefile(vim.split(text, "\n"), path, "b")
end

local function capture(rec, method, params, result)
  if method == "host/text" then
    copy_source(rec, params.file, result.text)
    return
  end
  local section = SECTION[method]
  if not section then
    return
  end
  local k = key(rec, params)
  if method == "host/references" then
    k = k .. (params.includeDeclaration and "|decl" or "|nodecl")
  end
  -- The engine's first positional request is at the traced identifier.
  rec.target = rec.target or { file = M._rel(params.file, rec.root), line = params.line, col = params.col }
  rec.recorded[section][k] = section == "documentHighlight" and vim.deepcopy(result) or relativised(rec, result)
end

function M.begin(dir, root)
  if active then
    error("whence: a recording into " .. active.dir .. " is already active")
  end
  vim.fn.mkdir(dir, "p")
  local rec = {
    dir = dir,
    root = root,
    recorded = { definition = {}, references = {}, documentHighlight = {} },
    orig = host.handle,
  }
  host.handle = function(method, params)
    local result, err = rec.orig(method, params)
    if not err and result ~= nil then
      capture(rec, method, params, result)
    end
    return result, err
  end
  active = rec
end

function M.finish()
  local rec = active
  if not rec then
    return nil
  end
  active = nil
  host.handle = rec.orig

  local sections = {}
  for name, entries in pairs(rec.recorded) do
    sections[name] = next(entries) and entries or vim.empty_dict()
  end
  vim.fn.writefile({ vim.json.encode(sections) }, rec.dir .. "/host.json")

  local meta = {
    root = rec.root,
    file = rec.target and rec.target.file or vim.NIL,
    line = rec.target and rec.target.line or vim.NIL,
    col = rec.target and rec.target.col or vim.NIL,
    engine_version = require("whence.version"),
    recorded_at = os.date("!%Y-%m-%dT%H:%M:%SZ"),
  }
  vim.fn.writefile({ vim.json.encode(meta) }, rec.dir .. "/whence-record.json")

  if not rec.target then
    return nil
  end
  return ("whence replay %s %s:%d:%d"):format(rec.dir, rec.target.file, rec.target.line + 1, rec.target.col + 1)
end

function M.run(dir, root)
  M.begin(dir, root)
  local finished = false
  local function done()
    if finished then
      return
    end
    finished = true
    local hint = M.finish()
    vim.notify("whence: " .. (hint or ("recorded nothing into " .. dir)))
  end
  local ok, err = pcall(require("whence").trace, done)
  if not ok then
    done()
    error(err)
  end
end

return M

File: vscode/src/record.ts
import { mkdir, readdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { host } from "./host";
import { fixtureKey, portablePath, relPath } from "./hostReplay";
import type { HostHandler, Location } from "./types";

const SECTION: Record<string, "definition" | "references" | "documentHighlight"> = {
  "host/definition": "definition",
  "host/references": "references",
  "host/documentHighlight": "documentHighlight",
};

type Recording = {
  dir: string;
  root: string;
  target: { file: string; line: number; col: number };
  engineVersion: string;
  recorded: { definition: Record<string, unknown>; references: Record<string, unknown>; documentHighlight: Record<string, unknown> };
  conflicts: string[];
  original: HostHandler;
};

let active: Recording | null = null;

async function copySource(rec: Recording, file: string, text: string): Promise<void> {
  const rel = relPath(file, rec.root);
  if (rel === null) {
    return;
  }
  const out = path.join(rec.dir, rel);
  await mkdir(path.dirname(out), { recursive: true });
  await writeFile(out, text);
}

async function capture(rec: Recording, method: string, params: any, result: unknown): Promise<void> {
  if (method === "host/text") {
    await copySource(rec, params.file, (result as { text: string }).text);
    return;
  }
  const section = SECTION[method];
  if (!section) {
    return;
  }
  const key = fixtureKey(rec.root, method, params);
  const answer =
    section === "documentHighlight"
      ? structuredClone(result)
      : (structuredClone(result) as Location[]).map((l) => ({ ...l, file: portablePath(l.file, rec.root) }));
  const first = rec.recorded[section][key];
  if (first !== undefined) {
    const label = `${method} ${key}`;
    if (JSON.stringify(first) !== JSON.stringify(answer) && !rec.conflicts.includes(label)) {
      rec.conflicts.push(label);
    }
    return;
  }
  rec.recorded[section][key] = answer;
}

export async function begin(opts: {
  dir: string;
  root: string;
  target: { file: string; line: number; col: number };
  engineVersion: string;
}): Promise<void> {
  if (active) {
    throw new Error(`a recording into ${active.dir} is already active`);
  }
  const rec: Recording = {
    ...opts,
    recorded: { definition: {}, references: {}, documentHighlight: {} },
    conflicts: [],
    original: host.handle,
  };
  active = rec; // Claimed before the first await, or two overlapping calls both pass the check.
  try {
    await mkdir(opts.dir, { recursive: true });
    if ((await readdir(opts.dir)).length > 0) {
      throw new Error(`${opts.dir} is not empty; record into a fresh directory`);
    }
  } catch (e) {
    active = null;
    throw e;
  }
  host.handle = async (method, params) => {
    const result = await rec.original(method, params);
    await capture(rec, method, params, result);
    return result;
  };
  active = rec;
}

export async function finish(): Promise<string[]> {
  const rec = active;
  if (!rec) {
    return [];
  }
  active = null;
  host.handle = rec.original;
  await writeFile(path.join(rec.dir, "host.json"), JSON.stringify(rec.recorded));
  const meta = {
    root: rec.root,
    file: relPath(rec.target.file, rec.root) ?? rec.target.file,
    line: rec.target.line,
    col: rec.target.col,
    engine_version: rec.engineVersion,
    recorded_at: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    conflicts: rec.conflicts,
  };
  await writeFile(path.join(rec.dir, "whence-record.json"), JSON.stringify(meta));
  return rec.conflicts;
}
